docs/design/validation.md

```diff
@@ -0,0 +1,1989 @@
+
+A new field that accepts a list of functions can be added to the `schema.Attribute` type. For example:
+
+```go
+schema.Attribute{
+    // ...
+    Validators: []T{
+        T,
+        T,
+    },
+}
+```
+
+In this case, the framework would perform the validation similar to the previous framework `All()` wrapper function. The logical `AND` type of value validation is overwhelmingly more common in practice, which will simplify provider implementations. This still allows for an `Any()` based wrapper (logical `OR`) to be inserted if necessary.
+
+Colocating the value validation behaviors in the schema definition, means it is easier for provider developers to discover this type of validation and correlate the validation logic to the name and type information. This proposal will feel familiar to existing provider developers. New provider developers will immediately know that multiple validations are supported.
+
+##### New Attribute With Value Validation Type(s)
+
+The `schema.Attribute` type could be converted to a Go interface type and split into capabilities, similar to other interface types in the framework. For example:
+
+```go
+type Attribute interface {
+    Type(context.Context) attr.Type
+    // ...
+}
+
+type AttributeWithValidators struct {
+    Attribute
+    Validators []T
+}
+
+// or more interfaces
+
+type AttributeWithValidators interface {
+    Attribute
+    Validators(/* ... */) []T
+}
+```
+
+This type of proposal, in isolation, feels extraneous given the current attribute implementation. The framework does not appear to benefit from this splitting and it seems desirable that all attributes should be able to enable value validation via optional data on the existing type.
+
+## Recommendations
+
+This section will summarize the proposals into specific recommendations for each topic. Code examples are provided in following sections to illustrate the concepts. The final section provides some future considerations for the framework and terraform-plugin-go.
+
+### Overview
+
+Defining all validation functionality via interface types will offer the framework the most flexibility for future enhancements while ensuring consistent implementations. The request and response pattern should be used to enable backwards (in the case of field deprecations) and forwards compatibility.
+
+All validation should be implemented separately from plan modifications as they address differing concerns and operations within the Terraform. Attribute validations should be implemented as a slice of the interface type on `schema.Attribute` while Data Source, Provider, and Resource level validation should be implemented as new extension interface types. Further helper functions and designs can reduce implementation details.
+
+### Data Source Example Implementation
+
+Example framework code:
+
+```go
+// ValidateDataSourceConfigRequest contains request information from the ValidateDataSourceConfig RPC.
+type ValidateDataSourceConfigRequest struct {
```

Comment:
Bikeshed: my super spicy take here is we can probably get away without including `Config` in all of these, making them `ValidateDataSourceRequest`, etc. just to cut down on the verbosity. I don't know that any semantics are lost. There's a danger of a new RPC being introduced in the future that makes things more ambiguous, but I think we could make _that_ specially named if that unlikely scenario occurs.

Reply:
Famous last words 😄  I think I will leave this for implementation consideration as to not confuse readers about naming differences from the RPCs, unless we want a naming section in _this_ documentation about why we strayed. Seems more important to capture in the commit message or a code comment.